packages/snaps-controllers/src/utils.ts

```diff
@@ -38,6 +40,38 @@ export function setDiff<
   ) as Diff<ObjectA, ObjectB>;
 }
 
+/**
+ * Calculate a difference between two permissions objects.
+ *
+ * Similar to `setDiff` except for one additional condition:
+ * Permissions in B should be removed from A if they exist in both and have differing caveats.
+ *
+ * @param permissionsA - An object containing one or more partial permissions.
+ * @param permissionsB - An object containing one or more partial permissions to be subtracted from A.
+ * @returns The permissions set A without properties from B.
+ */
+export function permissionsDiff<
+  PermissionsA extends Record<string, Pick<PermissionConstraint, 'caveats'>>,
+  PermissionsB extends Record<string, Pick<PermissionConstraint, 'caveats'>>,
+>(
+  permissionsA: PermissionsA,
+  permissionsB: PermissionsB,
+): Diff<PermissionsA, PermissionsB> {
+  return Object.entries(permissionsA).reduce<
+    Record<string, Pick<PermissionConstraint, 'caveats'>>
+  >((acc, [key, value]) => {
+    const isIncluded = key in permissionsB;
+    if (
+      !isIncluded ||
+      (isIncluded &&
+        !deepEqual(value.caveats ?? [], permissionsB[key].caveats ?? []))
```

Comment:
Just a note here, for my own benefit as much as anyone else:

`fast-deep-equal` is sensitive to array element order:
```
> const deepEqual = require('fast-deep-equal')
undefined
> deepEqual([1, 2, 3], [1, 2, 3])
true
> deepEqual([1, 2, 3], [3, 2, 1])
false
```

I don't believe that this is documented or enforced anywhere, but I believe that element order should _not_ be a consideration for caveat identity. In practice, I think it probably is, just like it will be here.

This problem should probably be solved in the PermissionController itself, where we'd want to add some kind of logic that always treats caveat value arrays as sets.

For present purposes, I don't think it matters, because we'll just revoke "identical" permissions only to immediately add them back.